Detectors/CTP/workflowLumi/src/RawToLumiConverterSpec.cxx

```diff
@@ -0,0 +1,183 @@
+{
+  mNTFToIntegrate = ctx.options().get<int>("NTF-to-IRaverage");
+}
+
+void RawToLumiConverterSpec::run(framework::ProcessingContext& ctx)
+{
+  const gbtword80_t bcidmask = 0xfff;
+  using InputSpec = o2::framework::InputSpec;
+  using ConcreteDataTypeMatcher = o2::framework::ConcreteDataTypeMatcher;
+  using Lifetime = o2::framework::Lifetime;
+  // mOutputHWErrors.clear();
+  o2::framework::DPLRawParser parser(ctx.inputs());
+  // setUpDummyLink
+  auto& inputs = ctx.inputs();
+  // if we see requested data type input with 0xDEADBEEF subspec and 0 payload this means that the "delayed message"
+  // mechanism created it in absence of real data from upstream. Processor should send empty output to not block the workflow
+  {
+    static size_t contDeadBeef = 0; // number of times 0xDEADBEEF was seen continuously
+    std::vector<InputSpec> dummy{InputSpec{"dummy", o2::framework::ConcreteDataMatcher{"CTP", "RAWDATA", 0xDEADBEEF}}};
+    for (const auto& ref : o2::framework::InputRecordWalker(inputs, dummy)) {
+      const auto dh = o2::framework::DataRefUtils::getHeader<o2::header::DataHeader*>(ref);
+      auto payloadSize = o2::framework::DataRefUtils::getPayloadSize(ref);
+      if (payloadSize == 0) {
+        auto maxWarn = o2::conf::VerbosityConfig::Instance().maxWarnDeadBeef;
+        if (++contDeadBeef <= maxWarn) {
+          LOGP(alarm, "Found input [{}/{}/{:#x}] TF#{} 1st_orbit:{} Payload {} : assuming no payload for all links in this TF{}",
+               dh->dataOrigin.str, dh->dataDescription.str, dh->subSpecification, dh->tfCounter, dh->firstTForbit, payloadSize,
+               contDeadBeef == maxWarn ? fmt::format(". {} such inputs in row received, stopping reporting", contDeadBeef) : "");
+        }
+        ctx.outputs().snapshot(o2::framework::Output{"CTP", "LUMI", 0, o2::framework::Lifetime::Timeframe}, mOutputLumiInfo);
+        return;
+      }
+    }
+    contDeadBeef = 0; // if good data, reset the counter
+  }
+  //
+  std::vector<lumiInfo> lumiPointsHBF1;
+  size_t countsMB = 0;
+  uint32_t payloadCTP;
+  uint32_t orbit0 = 0;
+  bool first = true;
+  gbtword80_t remnant = 0;
+  uint32_t size_gbt = 0;
+  for (auto it = parser.begin(); it != parser.end(); ++it) {
+    auto rdh = it.get_if<o2::header::RAWDataHeader>();
+    auto triggerOrbit = o2::raw::RDHUtils::getTriggerOrbit(rdh);
+
+    auto feeID = o2::raw::RDHUtils::getFEEID(rdh); // 0 = IR, 1 = TCR
+    auto linkCRU = (feeID & 0xf00) >> 8;
+    if (linkCRU == o2::ctp::GBTLinkIDIntRec) {
+      payloadCTP = o2::ctp::NIntRecPayload;
+    } else {
+      continue;
+    }
+    if (first) {
+      orbit0 = triggerOrbit;
+      first = false;
+    }
+    LOG(debug) << "RDH FEEid: " << feeID << " CTP CRU link:" << linkCRU << " Orbit:" << triggerOrbit;
+    // TF in 128 bits words
+    gsl::span<const uint8_t> payload(it.data(), it.size());
+    gbtword80_t gbtWord = 0;
+    int wordCount = 0;
+    std::vector<gbtword80_t> diglets;
+    if (orbit0 != triggerOrbit) {
```

Comment:
I am a little bit confused by the logic. In this `if` you are assigning the counts of the previous orbit, and aft the for loop below you are pushing to the lumiPointsHBF1 the counts of the last orbit, right? Why not simply counting for each orbit and assigning in the end of the `for` loop?

Reply:
Yes, I see. But if there is more than one page for HB then orbit will not change between RDHs ?

Reply:
No, multi-page HBs keep the same RDH, only page counter is incremented.

Reply:
@shahor02 : you did not comment here. I kept it as it is.

Reply:
sorry, comment on what?